File: execution/audit_event.py
"""
Auditoria do evento: status, números sorteados e duplicatas no ranking.

Uso:
    python audit_event.py
"""

import os
import json
from pathlib import Path
from collections import Counter
import httpx
from dotenv import load_dotenv

root = Path(__file__).parent.parent
load_dotenv(root / ".env.local")

URL = os.environ["NEXT_PUBLIC_SUPABASE_URL"].rstrip("/")
KEY = os.environ["SUPABASE_SERVICE_ROLE_KEY"]
H = {"apikey": KEY, "Authorization": f"Bearer {KEY}"}


def get(path, params=None):
    with httpx.Client(timeout=30) as c:
        r = c.get(f"{URL}/rest/v1/{path}", headers=H, params=params or {})
        r.raise_for_status()
        return r.json()


def dump(label, data):
    print(f"\n=== {label} ===")
    print(json.dumps(data, indent=2, default=str, ensure_ascii=False))


print("\n########## 1) EVENTS ##########")
events = get("events", {"select": "*", "order": "created_at.desc"})
dump("all events", events)

print("\n########## 2) BINGO_GAMES ##########")
games = get("bingo_games", {"select": "*", "order": "created_at.desc"})
dump("all games", games)

print("\n########## 3) BINGO_DRAWN_NUMBERS ##########")
drawn = get(
    "bingo_drawn_numbers",
    {"select": "*", "order": "drawn_at.desc", "limit": "500"},
)
print(f"total rows: {len(drawn)}")
if drawn:
    by_game = {}
    for d in drawn:
        by_game.setdefault(d["game_id"], []).append(d["number"])
    for gid, nums in by_game.items():
        print(f"\ngame_id={gid}  count={len(nums)}  unique={len(set(nums))}")
        dup = [n for n, c in Counter(nums).items() if c > 1]
        if dup:
            print(f"  duplicated numbers: {dup}")
        nums_sorted = sorted(set(nums))
        missing = [n for n in range(1, 76) if n not in set(nums)]
        print(f"  drawn (sorted): {nums_sorted}")
        print(f"  NOT drawn: {missing}")

print("\n########## 4) GAME_RESULTS ##########")
try:
    results = get("game_results", {"select": "*", "order": "created_at.desc"})
    dump("game_results", results)
except httpx.HTTPStatusError as e:
    print(f"no game_results: {e}")

print("\n########## 5) USERS (Denner?) ##########")
users = get(
    "users",
    {"select": "id,name,email,role", "order": "name.asc"},
)
denners = [u for u in users if "denner" in (u.get("name") or "").lower()
           or "denner" in (u.get("email") or "").lower()]
dump("denner matches", denners)
dump("total users", {"count": len(users)})

File: execution/audit_event.mjs
// Auditoria: status do evento, números sorteados, duplicatas no ranking.
// Executa com: node execution/audit_event.mjs

import { readFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const envPath = resolve(__dirname, "..", ".env.local");
const env = Object.fromEntries(
  readFileSync(envPath, "utf8")
    .split("\n")
    .filter(l => l.trim() && !l.startsWith("#"))
    .map(l => {
      const i = l.indexOf("=");
      return [l.slice(0, i).trim(), l.slice(i + 1).trim()];
    })
);

const URL = env.NEXT_PUBLIC_SUPABASE_URL.replace(/\/$/, "");
const KEY = env.SUPABASE_SERVICE_ROLE_KEY;
const H = { apikey: KEY, Authorization: `Bearer ${KEY}` };

async function get(path, params = {}) {
  const qs = new URLSearchParams(params).toString();
  const url = `${URL}/rest/v1/${path}${qs ? "?" + qs : ""}`;
  const r = await fetch(url, { headers: H });
  if (!r.ok) throw new Error(`${r.status} ${url}: ${await r.text()}`);
  return r.json();
}

function dump(label, data) {
  console.log(`\n=== ${label} ===`);
  console.log(JSON.stringify(data, null, 2));
}

console.log("\n########## 1) EVENTS ##########");
const events = await get("events", { select: "*", order: "created_at.desc" });
dump("all events", events);

console.log("\n########## 2) BINGO_GAMES ##########");
const games = await get("bingo_games", { select: "*", order: "created_at.desc" });
dump("all games", games);

console.log("\n########## 3) BINGO_DRAWN_NUMBERS ##########");
const drawn = await get("bingo_drawn_numbers", {
  select: "*",
  order: "drawn_at.desc",
  limit: "1000",
});
console.log(`total rows: ${drawn.length}`);
const byGame = {};
for (const d of drawn) (byGame[d.game_id] ??= []).push(d.number);
for (const [gid, nums] of Object.entries(byGame)) {
  const uniq = new Set(nums);
  const counts = {};
  for (const n of nums) counts[n] = (counts[n] ?? 0) + 1;
  const dup = Object.entries(counts).filter(([, c]) => c > 1);
  const missing = [];
  for (let n = 1; n <= 75; n++) if (!uniq.has(n)) missing.push(n);
  console.log(`\ngame_id=${gid}  rows=${nums.length}  unique=${uniq.size}`);
  if (dup.length) console.log(`  DUPLICATES: ${JSON.stringify(dup)}`);
  console.log(`  drawn (sorted): [${[...uniq].sort((a,b)=>a-b).join(",")}]`);
  console.log(`  NOT drawn: [${missing.join(",")}]`);
}

console.log("\n########## 4) GAME_RESULTS ##########");
try {
  const results = await get("game_results", { select: "*", order: "created_at.desc" });
  dump("game_results", results);
} catch (e) {
  console.log(`(no game_results or error): ${e.message}`);
}

console.log("\n########## 5) USERS (Denner?) ##########");
const users = await get("users", { select: "id,name,email,role", order: "name.asc" });
const denners = users.filter(u =>
  (u.name || "").toLowerCase().includes("denner") ||
  (u.email || "").toLowerCase().includes("denner")
);
dump("denner matches", denners);
console.log(`total users: ${users.length}`);
